### PR TITLE
chore: update azure/kubelogin to address CVE

### PR DESCRIPTION
**Pull request by imjasonh**
This PR attempts to update the project's dependency on `github.com/Azure/kubelogin` to the most recent version, which allows the project to update its dependency on `github.com/Azure/azure-sdk-for-go/sdk/azidentity` to a version that addresses a medium-severity CVE: https://github.com/advisories/GHSA-m5vv-6r4h-3vj9

I don't know whether this project is actually susceptible to this vulnerability, but it gets flagged for users of this project, and in general depending on very old releases of packages is not recommended -- this updates `kubelogin` from [v0.0.20 (Aug 2022)](https://github.com/Azure/kubelogin/releases/tag/v0.0.20) to [v0.1.4 (July this year)](https://github.com/Azure/kubelogin/releases/tag/v0.1.4)

The API has changed in breaking ways in the intervening ~2 years, and I assume this is part of the reason this old dependency has stuck around.

This is in part an update to https://github.com/argoproj/argo-cd/pull/16661 -- I'm not sure if that PR should just be updated or if that's in some permanently stuck state.

Checklist:

* [c] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates? **no**
* [x] I've updated documentation as required by this PR. **n/a**
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged. **n/a**
* [?] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines. **n/a**
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by imjasonh** [comment]
`build-only` CI failure seems to have failed flakily

```
ERROR: failed to solve: docker.io/library/node:23.0.0@sha256:e643c0b70dca9704dff42e12b17f5b719dbe4f95e6392fc2dfa0c5f02ea8044d: failed to resolve source metadata for docker.io/library/node:23.0.0@sha256:e643c0b70dca9704dff42e12b17f5b719dbe4f95e6392fc2dfa0c5f02ea8044d: failed to authorize: failed to fetch oauth token: unexpected status from POST request to https://auth.docker.io/token: 502 Bad Gateway

```

**Review by hectorj2f — APPROVED** [review]

**Comment by blakepettersson** [comment]
@imjasonh the unit test failure is a real one, which I can consistently reproduce.

What worked for me was reverting some of the unrelated changes (we would still need to find what the actual reason for the failure).

File: go.mod
Comment (blakepettersson) [inline go.mod:168]:
```suggestion
	github.com/Azure/go-ansiterm v0.0.0-20210617225240-d185dfc1b5a1 // indirect
```

File: go.mod
Comment (blakepettersson) [inline go.mod:193]:
```suggestion
	github.com/emicklei/go-restful/v3 v3.11.0 // indirect
```

File: go.mod
Comment (blakepettersson) [inline go.mod:199]:
```suggestion
	github.com/go-errors/errors v1.4.2 // indirect
```

File: go.mod
Comment (blakepettersson) [inline go.mod:287]:
```suggestion
	k8s.io/cli-runtime v0.31.0 // indirect
```

File: go.mod
Comment (blakepettersson) [inline go.mod:294]:
```suggestion
	sigs.k8s.io/json v0.0.0-20221116044647-bc3834ca7abd // indirect
	sigs.k8s.io/kustomize/api v0.17.2 // indirect
	sigs.k8s.io/kustomize/kyaml v0.17.1 // indirect
```

File: go.mod
Comment (blakepettersson) [inline go.mod:5]:
```suggestion
go 1.22.0
```

**Review by blakepettersson — COMMENTED** [review]
I added some changes to `go.mod` which makes the failing test pass (basically anything that's not directly related to `kubelogin` is reverted to its prior state for now), you'll then need to run `go mod tidy` afterwards.

**Comment by imjasonh** [comment]
Thanks @blakepettersson, I've updated the deps and ran `go mod tidy`, let me know if I missed anything.

**Comment by imjasonh** [comment]
Before this change (3625689264215d9e83bfafc5fe165c8fcd764ece):

```
govulncheck -show verbose ./...
Scanning your code and 2003 packages across 270 dependent modules for known vulnerabilities...

Fetching vulnerabilities from the database...

Checking the code against the vulnerabilities...

=== Symbol Results ===

No vulnerabilities found.

=== Package Results ===

Vulnerability #1: GO-2024-2918
    Azure Identity Libraries Elevation of Privilege Vulnerability in
    github.com/Azure/azure-sdk-for-go/sdk/azidentity
  More info: https://pkg.go.dev/vuln/GO-2024-2918
  Module: github.com/Azure/azure-sdk-for-go/sdk/azidentity
    Found in: github.com/Azure/azure-sdk-for-go/sdk/azidentity@v1.1.0
    Fixed in: github.com/Azure/azure-sdk-for-go/sdk/azidentity@v1.6.0
```

After this change (11ce4859db5dfe2c7d3fe00990464bbfc63afc8a):

```
govulncheck -show verbose ./...
Scanning your code and 2015 packages across 269 dependent modules for known vulnerabilities...

Fetching vulnerabilities from the database...

Checking the code against the vulnerabilities...

=== Symbol Results ===

No vulnerabilities found.

=== Package Results ===

No other vulnerabilities found.
```

**Comment by blakepettersson** [comment]
@imjasonh I'm no Azure expert - how has this been verified that this still works?

**Comment by imjasonh** [comment]
> @imjasonh I'm no Azure expert - how has this been verified that this still works?

I haven't verified anything at all. I'm just basing this on the previous PR from the feature's author to get the dep updated.

If you have any simple instructions to test this, or an end-to-end test that I could run, I'd be happy to give it a shot.

**Comment by imjasonh** [comment]
@bcho is the previous PR's author, 11 months ago

**Review by bcho — APPROVED** [review]
LGTM

**Comment by blakepettersson** [comment]
>If you have any simple instructions to test this, or an end-to-end test that I could run, I'd be happy to give it a shot.

Sadly I do not know of any. I guess given that @bcho is one of the actual authors of kubelogin and works for Azure Cloud that it gives some reassurance that this should work. I'd pass your question along to bcho.

**Comment by imjasonh** [comment]
Rebased and updated `go mod tidy`

**Comment by bcho** [comment]
Hi, thanks for the tag and the fix! - i am the contributor of kubelogin. I will try verify the change with an AKS cluster w/ entra id auth enabled tomorrow and report back here.

**Comment by bcho** [comment]
I have verified above code from a test env against an AAD + workload identity enabled AKS cluster. It worked after applying above changes for updating the backfill behavior for both devicelogin / msi / workload identity login method.

File: cmd/argocd-k8s-auth/commands/azure.go
Comment (imjasonh) [reply cmd/argocd-k8s-auth/commands/azure.go]: Thanks! I've applied this suggestion, please take a look.

**Review by blakepettersson — APPROVED** [review]
Thanks for your review, @bcho! LGTM

**Review by bcho — APPROVED** [review]
LGTM

**Comment by imjasonh** [comment]
Thanks @blakepettersson and @bcho! 

It looks like this still needs approval from an owner with write access. To avoid needing to keep this up to date (as `go.mod` is a fast-changing file), is there someone we could poke to get this merged soon? 🙏

**Review by rumstead — APPROVED** [review]

**Review by crenshaw-dev — APPROVED** [review]

**Comment by crenshaw-dev** [comment]
Since it's only medium severity, I don't plan to backport the fix. It'll go out with 2.14. Thanks all!

**Pull request merged**
